File: BasicLevel_C++/1031. 查验身份证(15).cpp
#include <iostream>
using namespace std;
int a[17] = {7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2};
int b[11] = {1, 0, 10, 9, 8, 7, 6, 5, 4, 3, 2};
string s;
bool isTrue() {
    int sum = 0;
    for (int i = 0; i < 17; i++) {
        if (s[i] < '0' || s[i] > '9') return false;
        sum += (s[i] - '0') * a[i];
    }
    int temp = (s[17] == 'X') ? 10 : (s[17] - '0');
    return b[sum%11] == temp;
}
int main() {
    int n, flag = 0;
    cin >> n;
    for (int i = 0; i < n; i++) {
        cin >> s;
        if (!isTrue()) {
            cout << s << endl;
            flag = 1;
        }
    }
    if (flag == 0) cout << "All passed";
    return 0;
}

File: BasicLevel_Golang/1031. 查验身份证(15).go
package main

import "fmt"

func isFake(id string) bool {
	var (
		sum       int
		weight    = [...]int{7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2}
		checkcode = [...]byte{'1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2'}
	)
	for i := 0; i < 17; i++ {
		if id[i] < '0' || id[i] > '9' {
			return true
		}
		sum += int(id[i]-'0') * weight[i]
	}
	z := sum % 11
	if checkcode[z] != id[17] {
		return true
	}
	return false
}

func main() {
	var (
		N, flag int
		id      string
	)
	fmt.Scanln(&N)
	for i := 0; i < N; i++ {
		fmt.Scanln(&id)
		if isFake(id) {
			fmt.Println(id)
			flag = 1
		}
	}
	if flag == 0 {
		fmt.Println("All passed")
	}
}
